### PR TITLE
Fix warning C4127: conditional expression is constant

### PR DESCRIPTION
**Pull request by antonioborondo**
https://github.com/nlohmann/json/issues/1268

* * *

## Pull request checklist

Read the [Contribution Guidelines](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md) for detailed information.

- [x]  Changes are described in the pull request, or an [existing issue is referenced](https://github.com/nlohmann/json/issues).
- [x]  The test suite [compiles and runs](https://github.com/nlohmann/json/blob/develop/README.md#execute-unit-tests) without error.
- [x]  [Code coverage](https://coveralls.io/github/nlohmann/json) is 100%. Test cases can be added by editing the [test suite](https://github.com/nlohmann/json/tree/develop/test/src).
- [x]  The source code is amalgamated; that is, after making changes to the sources in the `include/nlohmann` directory, run `make amalgamate` to create the single-header file `single_include/nlohmann/json.hpp`. The whole process is described [here](https://github.com/nlohmann/json/blob/develop/.github/CONTRIBUTING.md#files-to-change).

## Please don't

- The C++11 support varies between different **compilers** and versions. Please note the [list of supported compilers](https://github.com/nlohmann/json/blob/master/README.md#supported-compilers). Some compilers like GCC 4.7 (and earlier), Clang 3.3 (and earlier), or Microsoft Visual Studio 13.0 and earlier are known not to work due to missing or incomplete C++11 support. Please refrain from proposing changes that work around these compiler's limitations with `#ifdef`s or other means.
- Specifically, I am aware of compilation problems with **Microsoft Visual Studio** (there even is an [issue label](https://github.com/nlohmann/json/issues?utf8=✓&q=label%3A%22visual+studio%22+) for these kind of bugs). I understand that even in 2016, complete C++11 support isn't there yet. But please also understand that I do not want to drop features or uglify the code just to make Microsoft's sub-standard compiler happy. The past has shown that there are ways to express the functionality such that the code compiles with the most recent MSVC - unfortunately, this is not the main objective of the project.
- Please refrain from proposing changes that would **break [JSON](http://json.org) conformance**. If you propose a conformant extension of JSON to be supported by the library, please motivate this extension.
- Please do not open pull requests that address **multiple issues**.

### CODE REVIEW

**Comment by theodelrieu** [comment]
I was thinking of something like this (to not add runtime overhead):

```cpp
template <std::size_t N>
void fill_buffer() {
  fill_buffer_utf32();
}

template <>
void fill_buffer<2>() {
  fill_buffer_utf16();
}
```

Also, do not forget to edit files in `include/nlohmann` as well.

**Comment by antonioborondo** [comment]
@theodelrieu I just edited the include files and generated the header 👍 

About the implementation, I can change it if you want to use templates, but I am not sure how much overhead I am introducing with my implementation. I would say none actually 🤔

**Comment by theodelrieu** [comment]
Well, you are performing a runtime check (with the `if (sizeof(...))`), which should not silence the warning.

Unfortunately the templates way is the pre-C++17 way of doing `if constexpr`.

**Comment by nlohmann** [comment]
Yes, please use templates. The change right now makes little sense.

**Comment by antonioborondo** [comment]
@theodelrieu @nlohmann The warning is silenced with the current implementation, but I will use templates if you prefer.

**Comment by antonioborondo** [comment]
Commit https://github.com/nlohmann/json/pull/1272/commits/9ba3f79667baa3f1afacb09cdcf1cb0275b8f411 fixes the following error in [LGTM](https://lgtm.com/projects/g/nlohmann/json/logs/rev/pr-e90542ed789e9e5ea21ec641c976cb6c108f8c64/lang:cpp/stage:Build%20merge%20(863e5ce)):
```
[2018-10-03 10:49:06] [build] /opt/src/single_include/nlohmann/json.hpp:2026:14: error: explicit specialization in non-namespace scope ‘class nlohmann::detail::wide_string_input_adapter<WideStringType>’
[2018-10-03 10:49:06] [build]      template<>
[2018-10-03 10:49:06] [build]               ^
[2018-10-03 10:49:06] [build] /opt/src/single_include/nlohmann/json.hpp:2027:25: error: template-id ‘fill_buffer<2>’ in declaration of primary template
[2018-10-03 10:49:06] [build]      void fill_buffer<2>()
```

**Comment by antonioborondo** [comment]
Commit https://github.com/nlohmann/json/pull/1272/commits/7c385a4844bb9268163485d3e66d6a23f02b6815 fixes the following error in [LGTM](https://lgtm.com/projects/g/nlohmann/json/logs/rev/pr-210d2f72f9d81c9db97a7e44cbd33d5ee101b6aa/lang:cpp/stage:Build%20merge%20(c48455f)):
```
[2018-10-03 11:31:51] [build] /opt/src/single_include/nlohmann/json.hpp: In member function ‘void nlohmann::detail::wide_string_input_adapter<WideStringType>::fill_buffer()’:
[2018-10-03 11:31:51] [build] /opt/src/single_include/nlohmann/json.hpp:2023:9: error: ‘wide_string_input_helper’ was not declared in this scope
[2018-10-03 11:31:51] [build]          wide_string_input_helper<WideStringType, T>::fill_buffer(str, current_wchar, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
[2018-10-03 11:31:51] [build]          ^~~~~~~~~~~~~~~~~~~~~~~~
[2018-10-03 11:31:51] [build] /opt/src/single_include/nlohmann/json.hpp:2023:9: note: suggested alternative: ‘wide_string_input_adapter’
[2018-10-03 11:31:51] [build]          wide_string_input_helper<WideStringType, T>::fill_buffer(str, current_wchar, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
[2018-10-03 11:31:51] [build]          ^~~~~~~~~~~~~~~~~~~~~~~~
[2018-10-03 11:31:51] [build]          wide_string_input_adapter
[2018-10-03 11:31:51] [build] /opt/src/single_include/nlohmann/json.hpp:2023:48: error: expected primary-expression before ‘,’ token
[2018-10-03 11:31:51] [build]          wide_string_input_helper<WideStringType, T>::fill_buffer(str, current_wchar, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
[2018-10-03 11:31:51] [build]                                                 ^
[2018-10-03 11:31:51] [build] /opt/src/single_include/nlohmann/json.hpp:2023:54: error: ‘::fill_buffer’ has not been declared
[2018-10-03 11:31:51] [build]          wide_string_input_helper<WideStringType, T>::fill_buffer(str, current_wchar, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
[2018-10-03 11:31:51] [build]                                                       ^~~~~~~~~~~
[2018-10-03 11:31:52] [build] /opt/src/single_include/nlohmann/json.hpp:2023:54: note: suggested alternative: ‘setbuffer’
[2018-10-03 11:31:52] [build]          wide_string_input_helper<WideStringType, T>::fill_buffer(str, current_wchar, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
[2018-10-03 11:31:52] [build]                                                       ^~~~~~~~~~~
[2018-10-03 11:31:52] [build]                                                       setbuffer
```

File: include/nlohmann/detail/input/input_adapters.hpp [hunk 1]
```diff
@@ -115,144 +115,155 @@ class input_buffer_adapter : public input_adapter_protocol
     const char* const limit;
 };
 
-template<typename WideStringType>
-class wide_string_input_adapter : public input_adapter_protocol
+namespace
```
Comment (theodelrieu) [inline include/nlohmann/detail/input/input_adapters.hpp]: We're already in `namespace detail`, plus having anonymous namespaces in headers can be dangerous, as explained in #552

**Comment by nlohmann** [comment]
This pull request **fixes 1 alert** when merging 7c385a4844bb9268163485d3e66d6a23f02b6815 into 9f18e17063425db9d048a71d25ac3caaaa0738be - [view on LGTM.com](https://lgtm.com/projects/g/nlohmann/json/rev/pr-cd2dbb67fb741df0642c0bc13096417ca0e8ad4c)

**fixed alerts:**

* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by antonioborondo** [comment]
@theodelrieu PR ready to merge 👍

**Comment by nlohmann** [comment]
Thanks!

**Pull request merged**